pilot/pkg/networking/core/v1alpha3/route/route.go

```diff
@@ -535,27 +535,47 @@ func applyHTTPRouteDestination(
 		}
 	}
 
-	// TODO: eliminate this logic and use the total_weight option in envoy route
-	weighted := make([]*route.WeightedCluster_ClusterWeight, 0)
-	for _, dst := range in.Route {
-		weight := &wrappers.UInt32Value{Value: uint32(dst.Weight)}
-		if dst.Weight == 0 {
-			// Ignore 0 weighted clusters if there are other clusters in the route.
-			// But if this is the only cluster in the route, then add it as a cluster with weight 100
-			if len(in.Route) == 1 {
-				weight.Value = uint32(100)
-			} else {
+	if len(in.Route) == 1 {
+		processSingleDestination(in.Route[0], serviceRegistry, listenerPort, hashByDestination, out, action, false)
+	} else {
+		weighted := make([]*route.WeightedCluster_ClusterWeight, 0)
+		for _, dst := range in.Route {
+			if dst.Weight == 0 {
+				// Ignore 0 weighted clusters if there are other clusters in the route.
 				continue
 			}
+			weighted = append(weighted, processSingleDestination(dst, serviceRegistry, listenerPort, hashByDestination, out, action, true))
+		}
+		action.ClusterSpecifier = &route.RouteAction_WeightedClusters{
+			WeightedClusters: &route.WeightedCluster{
+				Clusters: weighted,
+			},
 		}
-		hostname := host.Name(dst.GetDestination().GetHost())
-		n := GetDestinationCluster(dst.Destination, serviceRegistry[hostname], listenerPort)
-		clusterWeight := &route.WeightedCluster_ClusterWeight{
+	}
+}
+
+func processSingleDestination(dst *networking.HTTPRouteDestination, serviceRegistry map[host.Name]*model.Service,
```

Comment:
can you split this for weighted cluster and single cluster

The processing is quite different

Reply:
The confusing point to me is the return value

Reply:
SGTM. Done